### ISSUE TITLE
API to explicitly label python input arguments in ``tf.function``

### ISSUE THREAD

**Issue by refraction-ray** [tensorflow/tensorflow#52193]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): tf 2.5.0
- Are you willing to contribute it (Yes/No): No

**Describe the feature and the current behavior/state.**

In Jax, ``tf.function``'s equivalent ``jax.jit`` has an argument ``static_argnums=(1, )``, which explicitly tells the compiler that the second(1) argument of the function to be jitted is a python object instead of a tensor, and the function should be retraced for each different argument in ``static_argnums``.

It seems to me that, ``tf.function`` has no corresponding API. Though, in general case, the function wrapped by ``tf.function`` can automatically determine whether the input is a tensor or a python object. Such automatic control fails when the function jitted is a function with ``tf.custom_gradient``. Please see the demo below.

```python
import tensorflow as tf
import numpy as np

class TensorWrapper:
    def __init__(self, tensor):
        self._tensor = tensor  # suppose we have no way to directly access the tensor

    def plus(self, other):
        # only public API
        return other + self._tensor

    def __str__(self):
        return "TensorWrapper(" + self._tensor.__str__() + ")"

    __repr__ = __str__


TW_list_tf = list(map(TensorWrapper, [tf.zeros([2, 2]), tf.ones([2, 2])]))

print(TW_list_tf)


def u_tf(a, b):
    return tf.reduce_sum(TW_list_tf[b].plus(a))

print("u_tf ", u_tf(tf.zeros([2, 2]), 1))

u_tf_jit = tf.function(u_tf)

print("u_tf_jit", u_tf_jit(tf.zeros([2, 2]), 1))

a = tf.zeros([2, 2])
with tf.GradientTape() as tape:
    tape.watch(a)
    loss = u_tf(a, 1)
print("u_tf grad ", tape.gradient(loss, a))

a = tf.zeros([2, 2])
with tf.GradientTape() as tape:
    tape.watch(a)
    loss = u_tf_jit(a, 1)
print("u_tf_jit grad ", loss, tape.gradient(loss, a))

@tf.custom_gradient
def u_tf_grad_v2(a, b):
    r = u_tf(a, b)

    def grad(dr):
        return 2.0 * dr * tf.ones_like(a), tf.zeros_like(b)

    return r, grad

a = tf.zeros([2, 2])
with tf.GradientTape() as tape:
    tape.watch(a)
    loss = u_tf_grad_v2(a, 1)
print("u_tf_grad_v2 grad ", tape.gradient(loss, a))

u_tf_grad_v2_jit = tf.function(u_tf_grad_v2)

try:
    a = tf.zeros([2, 2])
    with tf.GradientTape() as tape:
        tape.watch(a)
        loss = u_tf_grad_v2_jit(a, 1)
    print(loss, tape.gradient(loss, a))
except Exception as e:
    print("u_tf_grad_v2_jit grad:", e)
# list indices must be integers or slices, not Tensor since int b is compiled to tensor
```

There are two ways to solve the final exception. Add ``nondiff_argnums`` arguments in ``tf.custom_gradient`` API, or better add ``static_argnums`` arguments in ``tf.function`` API. Or is there currently any workaround for this ``tf.custom_gradient`` + ``tf.function`` + non tensor input issue?

**Will this change the current api? How?**
Yes, a new arguments ``static_argnums=`` should be added in ``tf.function()`` API.

**Who will benefit with this feature?**
Everyone wants better control on the jit behavior in tf and enjoys writting highly customized code with tf.

**Any Other info.**

**Comment by mohantym** [comment]
Hi @sanatmpa1!Could you please look at this feature request!

**Comment by refraction-ray** [comment]
@mdanatg , thanks for your reply. But the problem is that I want to label explicitly some arguments of the function as **non-Tensor** instead of ``tf.tensor``.

**Comment by mdanatg** [comment]
I see, what you seem to need is really a @custom_gradient with non-Tensor arguments. That seems to be supported when running eagerly, but inside tf.function, it seems to [convert everything to a tensor](https://github.com/tensorflow/tensorflow/blob/5d8492ab786724ed811588183c93b1623f02c3c5/tensorflow/python/ops/custom_gradient.py#L410). That looks like a bug. I'm also not sure if custom_gradient interacts well with tf.function. This is what is being done more commonly, and is what should work in your case:

```
@tf.custom_gradient
def u_tf_grad_v2(a, b):
    r = u_tf_jit(a, b)

    @tf.function
    def grad(dr):
        return 2.0 * dr * tf.ones_like(a), tf.zeros_like(b)

    return r, grad
```

That said, calling this custom gradient from within another tf.function won't work until the custom_gradient bug is fixed.

**Comment by sachinprasadhs** [comment]
@refraction-ray , If your issue is resolved with the above comment, could you please close the issue. For more details check https://www.tensorflow.org/guide/intro_to_graphs. Thanks!
